src/transaction/log_replication.cpp

```diff
@@ -36,10 +36,16 @@ namespace cublog
 {
   replicator::replicator (const log_lsa &start_redo_lsa)
     : m_redo_lsa { start_redo_lsa }
+    , m_perfmon_log_processing { PSTAT_SCAL_REPL_LOG_REDO_SYNC }
   {
     log_zip_realloc_if_needed (m_undo_unzip, LOGAREA_SIZE);
     log_zip_realloc_if_needed (m_redo_unzip, LOGAREA_SIZE);
 
+    // in case replication is applied synchronously, the perf subsystem needs
+    // to have a non-negative transation index on all threads
+    cubthread::entry &thread_entry = cubthread::get_entry ();
+    thread_entry.tran_index = LOG_SYSTEM_TRAN_INDEX;

```

Comment:
thread_entry.tran_index is not supposed to be set here. This is the thread that creates the replicator, probably the main thread during boot, or another thread.

The thread_entry that needs to be set to LOG_SYSTEM_TRAN_INDEX is the thread_entry of m_daemon. Give m_daemon a context manager that sets tran_index to LOG_SYSTEM_TRAN_INDEX.

Also, imo the replicator daemon thread should also be a TT_REPLICATION.

Reply:
You are so right. How did I miss that?